win-capture: Improve fullscreen window detection

Kira-NT opened the pull request:
### Description

This small change allows windows that extend slightly beyond the defined viewport to be properly detected as fullscreen ones.

While, in general, one should not assume that `bottom > top` and `right > left`, `win-capture` is tailored to Windows, where this assumption always holds true. Therefore, simply swapping `==` with `>=` when comparing `rect.right` with `mi.rcMonitor.right` and `rect.bottom` with `mi.rcMonitor.bottom` should suffice.

### Motivation and Context

While it may sound crazy, some apps on Windows use a dirty hack of extending their windows slightly beyond the monitor's declared resolution, usually by 1px, to preserve "true" windowed fullscreen and circumvent some issues related to the exclusive fullscreen mode presented by the graphics library being used, Windows itself, or both. As an example, see: https://github.com/glfw/glfw/issues/527#issuecomment-143183771.

At the moment, these windows are not recognized by OBS Game Capture because they fail the strict comparison rules defined here: https://github.com/obsproject/obs-studio/blob/48b1298fafb941f70edfa3b17bd2441e9124c3c9/plugins/win-capture/game-capture.c#L1221-L1228

And this is quite annoying because, from the user's perspective, these windows are clearly in the fullscreen mode.

Additionally, Windows itself doesn't resort to such strict comparison rules, as evidenced by the fact that the taskbar disappears, as it always does for fullscreen windows, even when a window stretches slightly beyond the defined viewport. So, if anything, this should be considered a parity feature.

### How Has This Been Tested?

There isn't much to test, so I just ran the freshly built OBS on Windows in a VM to check if it works. Thankfully, a two-symbol change didn't break the project :D

The change is very localized, shouldn't affect other areas of the code, and doesn't introduce any breaking changes. It simply fixes a minor issue/tweaks the pre-existing functionality to better match the, ahem, nuances of Windows development.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- Tweak (non-breaking change to improve existing functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation. *(I.e., none)*

Kira-NT committed:
win-capture: Improve fullscreen window detection

It's not that uncommon for apps on Windows to use a hacky method of
extending their windows ~1px past the supported monitor resolution
to preserve "true" windowed fullscreen and circumvent some exclusive
fullscreen-related issues presented by their graphics library,
Windows itself, or both.

Therefore, the fullscreen detection technique has been relaxed
to properly identify windows where `(x, y) == (viewportX, viewportY)`
and `(width, height) >= (viewportWidth, viewportHeight)` as fullscreen.

Windows also does something similar, evident by the fact that
the taskbar disappears, as it always does for fullscreen windows,
even when a window stretches slightly beyond the defined viewport.

derrod commented:
Please do not use merge commits, instead, rebase your PRs.

Kir-Antipov commented:
Apologies! It's been so long that I forgot what the contributing guidelines here look like, so I did the simplest thing possible - I resolved the conflict using GitHub's interface, since it was literally a one-line conflict, assuming the merge commit could be squashed away anyways :)

I've reordered the history, so the PR now consists of just one commit once again.

Warchamp7 commented on plugins/win-capture/game-capture.c:
```suggestion
static int rects_nearly_equal(RECT first, RECT second)
{
	return abs(first.left - second.left) <= 2 && abs(first.right - second.right) <= 2 &&
	       abs(first.top - second.top) <= 2 && abs(first.bottom - second.bottom) <= 2;
}
```

Warchamp7 commented on plugins/win-capture/game-capture.c:
```diff
@@ -1131,7 +1131,15 @@ static void get_fullscreen_window(struct game_capture *gc)
+	 * of a window against those of a monitor, and a more relaxed
+	 * one to check the bottom-right corner, allowing it to extend
+	 * slightly past the edge of the monitor */
+	if (rect.left == mi.rcMonitor.left && rect.right >= mi.rcMonitor.right && rect.bottom >= mi.rcMonitor.bottom &&
 	    rect.top == mi.rcMonitor.top) {
```
```suggestion
if (rects_nearly_equal(rect, mi.rcMonitor)) {
```

Warchamp7 commented on plugins/win-capture/game-capture.c:
```diff
@@ -1131,7 +1131,15 @@ static void get_fullscreen_window(struct game_capture *gc)
 	}
 
-	if (rect.left == mi.rcMonitor.left && rect.right == mi.rcMonitor.right && rect.bottom == mi.rcMonitor.bottom &&
+	/* some apps extend their windows ~1px beyond the monitor's
+	 * supported resolution to circumvent certain exclusive
+	 * fullscreen-related issues caused either by their graphics
+	 * library, Windows itself, or both; thus, we use a strict
+	 * comparison to check the coordinates of the top-left corner
+	 * of a window against those of a monitor, and a more relaxed
+	 * one to check the bottom-right corner, allowing it to extend
+	 * slightly past the edge of the monitor */
```
Comment needs to be adjusted for the suggested changes.

Warchamp7 reviewed (changes requested):
Tested with Helldivers 2
_(Which was annoying since its anticheat appears to prevent even opening a debugger with the game running)_

ShareX in screenshot mode will show the window as being 2560x1440 on my system, but a Print Screen results in a screenshot that it is 2562x1440.

Logging out the `GetWindowRect` from OBS shows the window rect as being -1, 2561, -1, 1441, which means it is reported by the Windows API as being 2562x1442.

This means that the solution in this PR of checking the top left corner explicitly would still fail, so I've proposed a +/- 2 px check on each corner instead.

I've suggested a 2px tolerance instead of 1px because testing with Slay the Spire 2 showed it also being 2562 wide, but the left of both rects is 0, and it is the right that is off by 2.

Warchamp7 commented:
@Kira-NT This is currently on our board for 33.0 but awaiting changes.

Kira-NT commented:
Thanks for the ping! Not sure how I managed to miss the previous one.

I've rebased the PR against the current master and adjusted the code as per your suggestions. I did add the acceptable delta as a parameter, though, instead of hardcoding it, so it's a bit easier to change in case in the future someone stumbles upon a game that does the same trick but with a 4px offset, for example, because after all my time researching this topic, I've never come across anyone suggesting a 2px offset instead of the 1px one *(although it does kinda make sense if you want to keep the window size a multiple of 2)*, so who knows what else might be out there in the wild.
